src/shims/sync.rs

```diff
@@ -392,197 +415,375 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+    }
+
+    fn pthread_cond_signal(&mut self, cond_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+        let id = cond_get_or_create_id(this, cond_op)?;
+        if let Some((thread, mutex)) = this.condvar_signal(id) {
+            post_cond_signal(this, thread, mutex)?;
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_cond_broadcast(&mut self, cond_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+        let id = cond_get_or_create_id(this, cond_op)?;
+
+        while let Some((thread, mutex)) = this.condvar_signal(id) {
+            post_cond_signal(this, thread, mutex)?;
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_cond_wait(
+        &mut self,
+        cond_op: OpTy<'tcx, Tag>,
+        mutex_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let id = cond_get_or_create_id(this, cond_op)?;
+        let mutex_id = mutex_get_or_create_id(this, mutex_op)?;
+        let active_thread = this.get_active_thread()?;
+
+        release_cond_mutex(this, active_thread, mutex_id)?;
+        this.condvar_wait(id, active_thread, mutex_id);
+
+        Ok(0)
+    }
+
+    fn pthread_cond_timedwait(
+        &mut self,
+        cond_op: OpTy<'tcx, Tag>,
+        mutex_op: OpTy<'tcx, Tag>,
+        abstime_op: OpTy<'tcx, Tag>,
+        dest: PlaceTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("pthread_cond_timedwait")?;
+
+        let id = cond_get_or_create_id(this, cond_op)?;
+        let mutex_id = mutex_get_or_create_id(this, mutex_op)?;
+        let active_thread = this.get_active_thread()?;
+
+        release_cond_mutex(this, active_thread, mutex_id)?;
+        this.condvar_wait(id, active_thread, mutex_id);
+
+        // We return success for now and override it in the timeout callback.
+        this.write_scalar(Scalar::from_i32(0), dest)?;
+
+        // Extract the timeout.
+        let clock_id = cond_get_clock_id(this, cond_op)?.to_i32()?;
+        let duration = {
+            let tp = this.deref_operand(abstime_op)?;
+            let mut offset = Size::from_bytes(0);
+            let layout = this.libc_ty_layout("time_t")?;
+            let seconds_place = tp.offset(offset, MemPlaceMeta::None, layout, this)?;
+            let seconds = this.read_scalar(seconds_place.into())?;
+            offset += layout.size;
+            let layout = this.libc_ty_layout("c_long")?;
+            let nanoseconds_place = tp.offset(offset, MemPlaceMeta::None, layout, this)?;
+            let nanoseconds = this.read_scalar(nanoseconds_place.into())?;
+            let (seconds, nanoseconds) = if this.pointer_size().bytes() == 8 {
+                (seconds.to_u64()?, nanoseconds.to_u64()?.try_into().unwrap())
+            } else {
+                (seconds.to_u32()?.into(), nanoseconds.to_u32()?)
+            };
```

Comment:
What about `(seconds.to_machine_usize(this)?, nanoseconds.to_machine_usize(this)?)`?